examples/pauseable.rs

```diff
@@ -0,0 +1,103 @@
+    // tracking how much time has passed. The spacebar key pauses/resumes.
+    //
+    // TODO: Produce input events (by wiggling the mouse, for example). More than 10 updates per
+    // second happen.
+
+    // 10 updates per second
+    let update_frequency = Duration::from_millis(100);
+    let started_at = Instant::now();
+
+    let mut running = true;
+    let mut last_update = Instant::now();
+    let mut total = Duration::new(0, 0);
+
+    event_loop.run(move |event, _, control_flow| {
+        match event {
+            Event::WindowEvent {
+                event: WindowEvent::CloseRequested,
+                ..
+            } => *control_flow = ControlFlow::Exit,
+            Event::WindowEvent {
+                event:
+                    WindowEvent::KeyboardInput {
+                        input:
+                            KeyboardInput {
+                                virtual_keycode: Some(VirtualKeyCode::Space),
+                                state: ElementState::Pressed,
+                                ..
+                            },
+                        ..
+                    },
+                ..
+            } => {
+                window.request_redraw();
+                running = !running;
+                if running {
+                    let now = Instant::now();
+                    last_update = now;
+                    *control_flow = ControlFlow::WaitUntil(now + update_frequency);
+                    println!("- Resuming!");
+                } else {
+                    *control_flow = ControlFlow::Wait;
+                    println!("- Pausing");
+                }
+            }
+            Event::MainEventsCleared => {
+                if running {
+                    let dt = last_update.elapsed();
+                    let now = Instant::now();
+                    // We want an update every update_frequency, so calculate the next time before
+                    // we process the update, which takes some amount of time. (If processing the
+                    // update takes more than update_frequency, then the next update will occur
+                    // immediately.)
+                    last_update = now;
+                    *control_flow = ControlFlow::WaitUntil(now + update_frequency);
+
+                    total += dt;
+                    println!(
+                        "- Update: +{:?} for a total of {:?}. It's been {:?} since the very start",
+                        dt,
+                        total,
+                        started_at.elapsed()
+                    );
+                    // Pretend that updating the application takes about 20ms
+                    thread::sleep(Duration::from_millis(20));
+                }
+                // Redraw since the application state was updated either by time passing or some
+                // input being handled previously.
+                window.request_redraw();
+            }
+            Event::WindowEvent { .. } => {
```

Comment:
That block should be moved to before `MainEventsCleared` ;)

I found this piece documentation useful : https://docs.rs/winit/0.21.0/winit/event/index.html
You should try to organize your code in the order the events are sent.